lib/util/request-batching.ts

```diff
@@ -0,0 +1,139 @@
+import * as errors from 'balena-errors';
+import chunk = require('lodash/chunk');
+import { groupByMap, mergePineOptions } from '.';
+import type {
+	PineOptionsWithSelect,
+	PineSelectableProps,
+	PineTypedResult,
+} from '..';
+
+const CHUNK_SIZE = 50;
+
+export function batchResourceOperationFactory<
+	T extends { id: number; uuid: string },
+>({
+	getAll,
+	NotFoundError,
+	AmbiguousResourceError,
+}: {
+	getAll: (options?: PineOptionsWithSelect<T>) => Promise<T[]>;
+	NotFoundError: new (id: string | number) => Error;
+	AmbiguousResourceError: new (id: string | number) => Error;
+}) {
+	type Item<TOpts> = { id: number } & (TOpts extends PineOptionsWithSelect<T>
+		? PineTypedResult<T, TOpts>
+		: {});
+
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>(options: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>) => Promise<void>;
+		groupByNavigationPoperty?: undefined;
+	}): Promise<void>;
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>(options: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>, ownerId: number) => Promise<void>;
+		groupByNavigationPoperty: PineSelectableProps<T>;
+	}): Promise<void>;
+	async function batchResourceOperation<
+		TOpts extends PineOptionsWithSelect<T>,
+	>({
+		uuidOrIdOrIds,
+		options,
+		groupByNavigationPoperty,
+		fn,
+	}: {
+		uuidOrIdOrIds: string | number | number[];
+		options?: TOpts;
+		fn: (items: Array<Item<TOpts>>, ownerId?: number) => Promise<void>;
+		groupByNavigationPoperty?: PineSelectableProps<T>;
+	}): Promise<void> {
+		if (
+			Array.isArray(uuidOrIdOrIds) &&
+			(!uuidOrIdOrIds.length ||
+				uuidOrIdOrIds.some((id) => typeof id !== 'number'))
+		) {
+			throw new errors.BalenaInvalidParameterError(
+				'uuidOrIdOrIds',
+				uuidOrIdOrIds,
+			);
+		}
+
+		// create a list of UUIDs or chunks of IDs
+		const chunks: Array<string | number[]> =
+			typeof uuidOrIdOrIds === 'string'
+				? [uuidOrIdOrIds]
+				: chunk(
+						!Array.isArray(uuidOrIdOrIds) ? [uuidOrIdOrIds] : uuidOrIdOrIds,
+						CHUNK_SIZE,
+				  );
```

Comment:
I think you can move the `!Array.isArray` outside of the chunks section, since it'll be an array of 1 anyway and you can't exactly chunk it any smaller:
```suggestion
			typeof uuidOrIdOrIds === 'string' || !Array.isArray(uuidOrIdOrIds)
				? [uuidOrIdOrIds]
				: chunk(
						 uuidOrIdOrIds,
						CHUNK_SIZE,
				  );
```

Reply:
Oh actually I see you're doing some magic around an array vs array of arrays below so you'd need
```suggestion
			typeof uuidOrIdOrIds === 'string'
				? [uuidOrIdOrIds]
				:  !Array.isArray(uuidOrIdOrIds)
				? [[uuidOrIdOrIds]]
				: chunk(
						 uuidOrIdOrIds,
						CHUNK_SIZE,
				  );
```
in which case maybe the existing version is better :/

Reply:
Exactly... I hope to untangle it a bit once I also add support for slugs.